File: app_express/src/repositories/dashboardRepository.js
const { assertNoError } = require('./base');
const { getCurrentBalance } = require('./paymentRepository');

function rangeQuery(query, column, start, end) {
  const endDate = new Date(`${end}T00:00:00`);
  endDate.setDate(endDate.getDate() + 1);
  return query.gte(column, start).lt(column, endDate.toISOString().slice(0, 10));
}

function groupBonsByFactory(bons) {
  const map = new Map();
  for (const row of bons) {
    const factoryId = row.factory_id || null;
    const name = row.factories?.name || 'Tanpa Pabrik';
    const key = factoryId || `__no__${name}`;
    const subTotal = (row.sub_notas || []).reduce((s, n) => s + Number(n.amount || 0), 0);
    let entry = map.get(key);
    if (!entry) {
      entry = { factory_id: factoryId, name, count: 0, tonnage: 0, value: 0 };
      map.set(key, entry);
    }
    entry.count += 1;
    entry.tonnage += Number(row.netto_2 || 0);
    entry.value += Number(row.total || 0) + subTotal;
  }
  return [...map.values()].sort((a, b) => b.tonnage - a.tonnage);
}

async function getDashboardStats(supabase, start, end) {
  const bonsQuery = rangeQuery(
    supabase.from('bons').select('id,status,factory_id,netto_2,total,factories(name),sub_notas(*)'),
    'bon_date',
    start,
    end
  );

  const running = await Promise.all([
    bonsQuery,
    rangeQuery(supabase.from('notas').select('id,status'), 'invoice_date', start, end),
    rangeQuery(supabase.from('payments').select('amount_paid'), 'payment_date', start, end),
    rangeQuery(supabase.from('margins').select('offtaker_amount'), 'transaction_date', start, end),
    rangeQuery(supabase.from('expenses').select('amount'), 'expense_date', start, end)
  ]).catch((error) => {
    // Fallback bila tabel sub_notas belum migrasi
    return Promise.all([
      rangeQuery(supabase.from('bons').select('id,status,factory_id,netto_2,total,factories(name)'), 'bon_date', start, end),
      rangeQuery(supabase.from('notas').select('id,status'), 'invoice_date', start, end),
      rangeQuery(supabase.from('payments').select('amount_paid'), 'payment_date', start, end),
      rangeQuery(supabase.from('margins').select('offtaker_amount'), 'transaction_date', start, end),
      rangeQuery(supabase.from('expenses').select('amount'), 'expense_date', start, end)
    ]);
  });

  const [bons, notas, payments, margins, expenses] = running;

  const bonsData = assertNoError(bons);
  const notasData = assertNoError(notas);
  const paymentsData = assertNoError(payments);
  const marginsData = assertNoError(margins);
  const expensesData = assertNoError(expenses);

  const totalMyTransactions = paymentsData.reduce((sum, row) => sum + Number(row.amount_paid || 0), 0);
  const totalTransactions = marginsData.reduce((sum, row) => sum + Number(row.offtaker_amount || 0), 0);
  const totalMargin = totalTransactions - totalMyTransactions;
  const totalExpenses = expensesData.reduce((sum, row) => sum + Number(row.amount || 0), 0);

  // Fokus transaksi harian
  const totalTonnage = bonsData.reduce((sum, row) => sum + Number(row.netto_2 || 0), 0);
  const totalTransactionValue = bonsData.reduce((sum, row) => {
    const subTotal = (row.sub_notas || []).reduce((s, n) => s + Number(n.amount || 0), 0);
    return sum + Number(row.total || 0) + subTotal;
  }, 0);
  const factoryBreakdown = groupBonsByFactory(bonsData);

  return {
    totalBons: bonsData.length,
    ongoingBons: bonsData.filter((row) => row.status !== 'LUNAS').length,
    finishedBons: bonsData.filter((row) => row.status === 'LUNAS').length,
    unpaidNotas: notasData.filter((row) => row.status !== 'LUNAS').length,
    paidNotas: notasData.filter((row) => row.status === 'LUNAS').length,
    totalTonnage,
    totalTransactionValue,
    totalPayments: totalMyTransactions,
    factoryBreakdown,
    totalTransactions,
    totalMyTransactions,
    totalMargin,
    totalExpenses,
    netProfit: totalMargin - totalExpenses,
    currentBalance: await getCurrentBalance(supabase)
  };
}

module.exports = { getDashboardStats };

File: lib/repositories/dashboard_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';

class FactoryBreakdown {
  final String? factoryId;
  final String name;
  final int count;
  final double tonnage;
  final double value;

  FactoryBreakdown({
    this.factoryId,
    required this.name,
    required this.count,
    required this.tonnage,
    required this.value,
  });
}

class DashboardStats {
  final int totalBons;
  final int ongoingBons;
  final int finishedBons;
  final int unpaidNotas;
  final int paidNotas;
  final double totalMargin;
  final double totalTransactions;
  final double totalMyTransactions;
  final double totalExpenses;
  final double netProfit;
  final int currentBalance;

  // Fokus transaksi harian
  final double totalTonnage;
  final double totalTransactionValue;
  final double totalPayments;
  final List<FactoryBreakdown> factoryBreakdown;

  DashboardStats({
    required this.totalBons,
    required this.ongoingBons,
    required this.finishedBons,
    required this.unpaidNotas,
    required this.paidNotas,
    required this.totalMargin,
    required this.totalTransactions,
    required this.totalMyTransactions,
    required this.totalExpenses,
    required this.netProfit,
    required this.currentBalance,
    this.totalTonnage = 0,
    this.totalTransactionValue = 0,
    this.totalPayments = 0,
    this.factoryBreakdown = const [],
  });
}

class DashboardRepository {
  final SupabaseClient _client;

  DashboardRepository(this._client);

  Future<DashboardStats> getDashboardStats(DateTime start, DateTime end) async {
    final startIso = start.toIso8601String();
    final nextDayEndIso = DateTime.utc(end.year, end.month, end.day + 1).toIso8601String();

    // 1. Bons Stats (termasuk netto_2 + total + factory name untuk breakdown)
    final bonsResponse = await _client
        .from('bons')
        .select('id, status, factory_id, netto_2, total, factories(name)')
        .gte('bon_date', startIso)
        .lt('bon_date', nextDayEndIso);

    final bonsList = (bonsResponse as List).cast<Map<String, dynamic>>();
    final totalBons = bonsList.length;
    final ongoingBons = bonsList.where((b) => b['status'] != 'LUNAS').length;
    final finishedBons = bonsList.where((b) => b['status'] == 'LUNAS').length;

    double totalTonnage = 0;
    double totalTransactionValue = 0;
    final factoryMap = <String, FactoryBreakdown>{};
    for (final b in bonsList) {
      final netto = (b['netto_2'] as num?)?.toDouble() ?? 0.0;
      final value = (b['total'] as num?)?.toDouble() ?? 0.0;
      totalTonnage += netto;
      totalTransactionValue += value;
      final factoryId = b['factory_id']?.toString();
      final factories = b['factories'];
      final name = (factories is Map && factories.containsKey('name'))
          ? (factories['name']?.toString() ?? 'Tanpa Pabrik')
          : 'Tanpa Pabrik';
      final id = factoryId ?? '$name-';
      final entry = factoryMap.putIfAbsent(
        id,
        () => FactoryBreakdown(
          factoryId: factoryId,
          name: name,
          count: 0,
          tonnage: 0,
          value: 0,
        ),
      );
      factoryMap[id] = FactoryBreakdown(
        factoryId: factoryId,
        name: name,
        count: entry.count + 1,
        tonnage: entry.tonnage + netto,
        value: entry.value + value,
      );
    }
    final factoryBreakdown = factoryMap.values.toList()
      ..sort((a, b) => b.tonnage.compareTo(a.tonnage));

    // 2. Nota Timbangan Stats
    final notasResponse = await _client
        .from('notas')
        .select('id, status')
        .gte('invoice_date', startIso)
        .lt('invoice_date', nextDayEndIso);

    final notasList = (notasResponse as List).cast<Map<String, dynamic>>();
    final unpaidNotas = notasList.where((i) => i['status'] != 'LUNAS').length;
    final paidNotas = notasList.where((i) => i['status'] == 'LUNAS').length;

    // 3. Financial Stats (Payments & Margins)
    final paymentsResponse = await _client
        .from('payments')
        .select('amount_paid')
        .gte('payment_date', startIso)
        .lt('payment_date', nextDayEndIso);

    double totalPayments = 0;
    for (var p in (paymentsResponse as List)) {
      totalPayments += (p['amount_paid'] as num?)?.toDouble() ?? 0.0;
    }

    final marginsResponse = await _client
        .from('margins')
        .select('offtaker_amount')
        .gte('transaction_date', startIso)
        .lt('transaction_date', nextDayEndIso);

    double totalTransactions = 0;
    for (var m in (marginsResponse as List)) {
      totalTransactions += (m['offtaker_amount'] as num?)?.toDouble() ?? 0.0;
    }

    final double totalMargin = totalTransactions - totalPayments;

    // 4. Expense Stats
    final expensesResponse = await _client
        .from('expenses')
        .select('amount')
        .gte('expense_date', startIso)
        .lt('expense_date', nextDayEndIso);

    double totalExpenses = 0;
    for (var e in (expensesResponse as List)) {
      totalExpenses += (e['amount'] as num?)?.toDouble() ?? 0.0;
    }

    final double netProfit = totalMargin - totalExpenses;

    // 5. Global Balance
    final allDepositsResponse = await _client.from('deposits').select('amount');
    int totalDeposits = 0;
    for (var d in (allDepositsResponse as List)) {
      totalDeposits += (d['amount'] as int);
    }

    final allPaymentsResponse = await _client
        .from('payments')
        .select('amount_paid');
    int totalPaymentsGlobal = 0;
    for (var p in (allPaymentsResponse as List)) {
      totalPaymentsGlobal += (p['amount_paid'] as int);
    }

    final int currentBalance = totalDeposits - totalPaymentsGlobal;

    return DashboardStats(
      totalBons: totalBons,
      ongoingBons: ongoingBons,
      finishedBons: finishedBons,
      unpaidNotas: unpaidNotas,
      paidNotas: paidNotas,
      totalMargin: totalMargin,
      totalTransactions: totalTransactions,
      totalMyTransactions: totalPayments,
      totalExpenses: totalExpenses,
      netProfit: netProfit,
      currentBalance: currentBalance,
      totalTonnage: totalTonnage,
      totalTransactionValue: totalTransactionValue,
      totalPayments: totalPayments,
      factoryBreakdown: factoryBreakdown,
    );
  }
}